substrate/frame/assets-vesting/src/lib.rs

```diff
@@ -0,0 +1,775 @@
+		///
+		/// - `asset`: Id of the asset class of the asset account for which the vesting applies.
+		/// - `target`: The account receiving the vested funds.
+		/// - `schedule`: The vesting schedule attached to the transfer.
+		///
+		/// Emits `VestingCreated`.
+		///
+		/// NOTE: This will unlock all schedules through the current block.
+		///
+		/// ## Complexity
+		/// - `O(1)`.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::vested_transfer(T::MAX_VESTING_SCHEDULES))]
+		pub fn vested_transfer(
+			origin: OriginFor<T>,
+			asset: AssetIdOf<T, I>,
+			target: AccountIdLookupOf<T>,
+			schedule: VestingInfo<BalanceOf<T, I>, BlockNumberFor<T>>,
+		) -> DispatchResult {
+			let transactor = ensure_signed(origin)?;
+			let target = T::Lookup::lookup(target)?;
+			Self::do_vested_transfer(asset, &transactor, &target, schedule)
+		}
+
+		/// Force a vested transfer.
+		///
+		/// The dispatch origin for this call must be `ForceOrigin`.
+		///
+		/// - `asset`: Id of the asset class of the asset account for which the vesting applies.
+		/// - `source`: The account whose funds should be transferred.
+		/// - `target`: The account that should be transferred the vested funds.
+		/// - `schedule`: The vesting schedule attached to the transfer.
+		///
+		/// Emits `VestingCreated`.
+		///
+		/// NOTE: This will unlock all schedules through the current block.
+		///
+		/// ## Complexity
+		/// - `O(1)`.
+		#[pallet::call_index(3)]
+		#[pallet::weight(T::WeightInfo::force_vested_transfer(T::MAX_VESTING_SCHEDULES))]
+		pub fn force_vested_transfer(
+			origin: OriginFor<T>,
+			asset: AssetIdOf<T, I>,
+			source: AccountIdLookupOf<T>,
+			target: AccountIdLookupOf<T>,
+			schedule: VestingInfo<BalanceOf<T, I>, BlockNumberFor<T>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let target = T::Lookup::lookup(target)?;
+			let source = T::Lookup::lookup(source)?;
+			Self::do_vested_transfer(asset, &source, &target, schedule)
+		}
+
+		/// Merge two vesting schedules together, creating a new vesting schedule that unlocks over
+		/// the highest possible start and end blocks. If both schedules have already started the
+		/// current block will be used as the schedule start; with the caveat that if one schedule
+		/// is finished by the current block, the other will be treated as the new merged schedule,
+		/// unmodified.
+		///
+		/// NOTE: If `schedule1_index == schedule2_index` this is a no-op.
+		/// NOTE: This will unlock all schedules through the current block prior to merging.
+		/// NOTE: If both schedules have ended by the current block, no new schedule will be created
+		/// and both will be removed.
+		///
+		/// Merged schedule attributes:
+		/// - `starting_block`: `MAX(schedule1.starting_block, scheduled2.starting_block,
+		///   current_block)`.
+		/// - `ending_block`: `MAX(schedule1.ending_block, schedule2.ending_block)`.
+		/// - `locked`: `schedule1.locked_at(current_block) + schedule2.locked_at(current_block)`.
```

Comment:
isn't it disadvantageous to merge if 1 or 2 schedule are future?
all the unlock from the earlier to the later is postponed to after later?

Hmm something seems wrong, if I have 2 identical vesting schedule, and I merge them I get a new vesting schedule that double the frozen amount.

AFAICT it is never advantageous to merge schedules, in any circumstances no? Do I miss something, why do have this feature to merge schedules?

Reply:
Hmm… good question! It was already in the `Currency` version (`pallet-vesting`), so I just moved it here to keep compatibility. Maybe we don't really need it, but there might be use cases that already take advantage from it. Not sure whether those assumptions would hold.

Reply:
Ok, let's keep it as it is, but let's add in the doc of the call that the resulting vesting is always disadvantageous to the caller, and the only advantage is to reduce the number of vesting schedule in storage.